bd-client-stats/src/file_manager.rs

```diff
@@ -0,0 +1,424 @@
+        <= self.time_provider.now()
+      {
+        log::debug!("snapshot is too old, creating new snapshot");
+        true
+      } else {
+        false
+      }
+    }) {
+      if self.max_aggregated_files.read() <= u32::try_from(initialized_inner.index.len()).unwrap() {
+        log::debug!("max files reached, popping oldest snapshot");
+        initialized_inner.delete_oldest_snapshot().await?;
+      }
+
+      let pending_file = PendingFile {
+        name: TrackedStatsUploadRequest::upload_uuid(),
+        period_start: self.time_provider.now().into_proto(),
+        ready_to_upload: false,
+        ..Default::default()
+      };
+      log::debug!("creating new snapshot in index: {}", pending_file.name);
+      initialized_inner.index.push_back(pending_file);
+      initialized_inner.write_index().await?;
+      false
+    } else {
+      true
+    };
+
+    // Read the file back or make a new one. We don't count an error reading the file or file
+    // corruption as a fatal error.
+    let path = STATS_DIRECTORY.join(&initialized_inner.index.back().unwrap().name);
+    let stats_upload_request = if use_existing {
+      initialized_inner
+        .file_system
+        .read_file(&path)
+        .await
+        .and_then(|contents| read_compressed_protobuf::<StatsUploadRequest>(&contents))
+        .inspect_err(|e| {
+          log::debug!(
+            "unable to read snapshot {}, creating default: {e}",
+            path.display()
+          );
+        })
+        .ok()
+    } else {
+      None
+    }
+    .unwrap_or_else(|| StatsUploadRequest {
+      upload_uuid: initialized_inner.index.back().unwrap().name.clone(),
+      ..Default::default()
+    });
+
+    Ok(StatsUploadRequestHandle {
+      index: initialized_inner.index.len() - 1,
+      stats_upload_request,
+    })
+  }
+
+  // Called if a merge results in no metrics. The file is removed from the index as it is never
+  // written to avoid empty uploads.
+  pub async fn remove_empty_snapshot(&self) -> anyhow::Result<()> {
+    let mut inner = self.inner.lock().await;
+    let initialized_inner = inner.get_initialized().await?;
+
+    debug_assert!(!initialized_inner.index.back().unwrap().ready_to_upload);
+    log::debug!(
+      "removing empty snapshot from index: {}",
+      initialized_inner.index.back().unwrap().name
+    );
+    initialized_inner.index.pop_back();
+    initialized_inner.write_index().await?;
+
+    Ok(())
+  }
+
+  // Write the snapshot returned from `get_or_create_snapshot` back to disk.
+  pub async fn write_snapshot(
+    &self,
+    mut handle: StatsUploadRequestHandle,
+    mut snapshot: Snapshot,
+  ) -> anyhow::Result<()> {
+    let mut inner = self.inner.lock().await;
+    let initialized_inner = inner.get_initialized().await?;
+
+    // Even for a pre-existing snapshot, we are being passed back a new snapshot so we need to
+    // initialized the period start again.
+    snapshot.occurred_at = Some(Occurred_at::Aggregated(Aggregated {
```

Comment:
I wonder if we should be storing both start/end of the snapshot and send both? Especially with the new jitter I think we'll end up with even more variable snapshot periods so maybe the server should insert them at avg(start, end) instead of always using the start?

Reply:
See above re: jitter, but having end makes sense to me. I can add that to the proto.